src/main/java/org/opensearch/security/privileges/RestLayerPrivilegesEvaluator.java

```diff
@@ -0,0 +1,119 @@
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.security.privileges;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.greenrobot.eventbus.Subscribe;
+import org.opensearch.OpenSearchSecurityException;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.transport.TransportAddress;
+import org.opensearch.common.util.concurrent.ThreadContext;
+import org.opensearch.core.xcontent.NamedXContentRegistry;
+import org.opensearch.security.auditlog.AuditLog;
+import org.opensearch.security.configuration.ClusterInfoHolder;
+import org.opensearch.security.configuration.ConfigurationRepository;
+import org.opensearch.security.securityconf.ConfigModel;
+import org.opensearch.security.securityconf.DynamicConfigModel;
+import org.opensearch.security.securityconf.SecurityRoles;
+import org.opensearch.security.support.ConfigConstants;
+import org.opensearch.security.user.User;
+import org.opensearch.threadpool.ThreadPool;
+
+import java.util.Set;
+
+public class RestLayerPrivilegesEvaluator {
+    protected final Logger log = LogManager.getLogger(this.getClass());
+    private final ClusterService clusterService;
+
+    private final AuditLog auditLog;
+    private ThreadContext threadContext;
+
+    private final ClusterInfoHolder clusterInfoHolder;
+    private ConfigModel configModel;
+    private DynamicConfigModel dcm;
+    private final NamedXContentRegistry namedXContentRegistry;
+
+    public RestLayerPrivilegesEvaluator(final ClusterService clusterService, final ThreadPool threadPool,
+                               final ConfigurationRepository configurationRepository,
+                               AuditLog auditLog, final Settings settings, final ClusterInfoHolder clusterInfoHolder,
+                               NamedXContentRegistry namedXContentRegistry) {
+
+        super();
+        this.clusterService = clusterService;
+        this.auditLog = auditLog;
+
+        this.threadContext = threadPool.getThreadContext();
+
+        this.clusterInfoHolder = clusterInfoHolder;
+        this.namedXContentRegistry = namedXContentRegistry;
+    }
+
+    @Subscribe
+    public void onConfigModelChanged(ConfigModel configModel) {
+        this.configModel = configModel;
+    }
+
+    @Subscribe
+    public void onDynamicConfigModelChanged(DynamicConfigModel dcm) {
+        this.dcm = dcm;
+    }
+
+    private SecurityRoles getSecurityRoles(Set<String> roles) {
+        return configModel.getSecurityRoles().filter(roles);
+    }
+
+    public boolean isInitialized() {
+        return configModel !=null && configModel.getSecurityRoles() != null && dcm != null;
+    }
+
+    public PrivilegesEvaluatorResponse evaluate(final User user, String action0) {
```

Comment:
This may be hit further up in the call chain. For instance the plugin itself probably cannot get the request to this point if security is not initialized.

Reply:
That's exactly how this would work. This evaluator is used in the REST handler wrapper which means this is executed before the original handler is called. If the request is not authorized, the original handler will never be called.